client/recordUtil.js

```diff
@@ -132,151 +28,62 @@ const normalizeBookmark = (record) => {
- * objectData to the final object that should be applied by the browser.
- * @param {Object} record SyncRecord JS object
- * @param {Object=} existingObject Browser object as syncRecord JS object
- * @returns {Object|null} Resolved syncRecord to apply to browser data
+ * Keeps only most the right pairs in array for each unique objectId from recordsAndExistingObjects
  */
-const resolveRecordWithObject = (record, existingObject) => {
-  const type = record.objectData
-  if (type === 'siteSetting') {
-    return resolveSiteSettingsRecordWithObject(record, existingObject)
-  }
-  if (record.action === proto.actions.UPDATE) {
-    if (deepEqual(record[type], existingObject[type])) {
-      // no-op
-      return null
+module.exports.keepMostRight = (recordsAndExistingObjects) => {
+  var alreadySeenObjectId = new Set()
+  for (var i = recordsAndExistingObjects.length - 1; i >= 0; --i) {
+    if (recordsAndExistingObjects[i][0]) {
+      const stringObjectId = `${recordsAndExistingObjects[i][0].objectId}`
+      if (alreadySeenObjectId.has(stringObjectId)) {
+        recordsAndExistingObjects.splice(i, 1)
+      } else {
+        alreadySeenObjectId.add(stringObjectId)
+      }
     }
-    return record
-  } else if (record.action === proto.actions.DELETE) {
-    return record
-  } else {
-    throw new Error('Invalid record action')
   }
+  return recordsAndExistingObjects
 }
 
 /**
- * Given a siteSettings SyncRecord and a browser's matching existing object, resolve
- * objectData to only have the applicable fields.
- * TODO: Maybe make behavior for siteSettings same as for other types.
- * @param {Object} record SyncRecord JS object
- * @param {Object=} existingObject Browser object as syncRecord JS object
- * @returns {Object|null} Resolved syncRecord to apply to browser data
+ * Gets the record from if its timestamp greater, or action is DELETE
+ * If any of the records does not have the timestamp, choose server record
  */
-const resolveSiteSettingsRecordWithObject = (record, existingObject) => {
-  const commonFields = ['hostPattern']
-  const type = record.objectData
-  const recordFields = new Set(Object.keys(record[type]))
-  const existingFields = new Set(Object.keys(existingObject[type]))
+module.exports.getThisPairWinner = (recordAndLocalObject) => {
+  const record = recordAndLocalObject[0]
+  const object = recordAndLocalObject[1]
 
-  let resolveField = null
-  if (record.action === proto.actions.UPDATE) {
-    resolveField = (field) => {
-      return !commonFields.includes(field) &&
-      (!existingFields.has(field) || !deepEqual(existingObject[type][field], record[type][field]))
-    }
-  } else if (record.action === proto.actions.DELETE) {
-    resolveField = (field) => {
-      return !commonFields.includes(field) && existingFields.has(field)
-    }
+  if (record.action === proto.actions.DELETE ||
+    !object) {
+    return record
+  } else if (object.action === proto.actions.DELETE) {
```

Comment:
why do we need to pull out these two DELETE cases?

Reply:
I think, DELETE is a special case of the operation. When it happens, the object is destroyed and no more sense  to perform any next operation on the object.

I chose `DELETE wins` way after considering this:
1. deviceA and deviceB both has synced bookmark `bookmarkA` 
2. at almost the same time `bookmarkA` is renamed on deviceA and removed on deviceB
3. what would I expect?

I would expect  `bookmarkA` will be removed on both devices, so I put this into the code.

And what do you think?

Reply:
we still need to consider timestamp because Android and brave-core now save object id even if the bookmark is deleted, which is the measure for preventing duplicate bookmarks when leaving and rejoining the sync chain. So DELETE still need to compete with CREATE or even UPDATE which is your case .